pkg/apis/migration/migration.go

```diff
@@ -92,46 +92,46 @@ const (
 	CRDPlural = "cnsvspherevolumemigrations"
 )
 
-// onceForVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
-var onceForVolumeMigrationService sync.Once
-
-// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
-var volumeMigrationInstance *volumeMigration
-
-var volumeMigrationServiceInitErr error
+var (
+	// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+	volumeMigrationInstance *volumeMigration
+	// volumeMigrationInstanceLock is used for handling race conditions during read, write on volumeMigrationInstance
+	volumeMigrationInstanceLock sync.Mutex
+)
 
 // GetVolumeMigrationService returns the singleton VolumeMigrationService
 func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config) (VolumeMigrationService, error) {
 	log := logger.GetLogger(ctx)
-	onceForVolumeMigrationService.Do(func() {
+	volumeMigrationInstanceLock.Lock()
+	defer volumeMigrationInstanceLock.Unlock()
+	if volumeMigrationInstance != nil {
```

Comment:
If the `GetVolumeMigrationService` fails first time, we might end up returning partially initialized `volumeMigrationInstance` right?

Reply:
Will the latest commit address your concern ?

Reply:
To make this singleton more efficient, can we use double checked locking? See https://launchdarkly.com/blog/golang-pearl-thread-safe-writes-and-double-checked-locking-in-go/